src/irmin-pack/traverse_pack_file.ml

```diff
@@ -56,7 +61,65 @@ end = struct
      as requests for more data. *)
   exception Not_enough_buffer
 
-  type index_binding = { key : Hash.t; data : Index.value }
+  type index_value = int63 * int * Pack_value.Kind.t
+  [@@deriving irmin ~equal ~pp]
+
+  type index_binding = { key : Hash.t; data : index_value }
+
+  module Index_reconstructor = struct
+    let create ?output config =
+      let dest =
+        match output with
+        | Some path -> path
+        | None ->
+            if Conf.readonly config then raise S.RO_not_allowed;
+            Conf.root config
+      in
+      let log_size = Conf.index_log_size config in
+      Log.app (fun f ->
+          f "Beginning index reconstruction with parameters: { log_size = %d }"
+            log_size);
+      let index = Index.v ~fresh:true ~readonly:false ~log_size dest in
+      index
+
+    let iter_pack_entry index key data = Index.add index key data
+
+    let finalise index () =
+      (* Ensure that the log file is empty, so that subsequent opens with a
+         smaller [log_size] don't immediately trigger a merge operation. *)
+      Log.app (fun f ->
+          f "Completed indexing of pack entries. Running a final merge ...");
+      Index.try_merge index;
+      Index.close index
+  end
+
+  module Index_checker = struct
+    let create config =
+      let log_size = Conf.index_log_size config in
+      Log.app (fun f ->
+          f "Beginning index checking with parameters: { log_size = %d }"
+            log_size);
+      let index =
+        Index.v ~fresh:false ~readonly:true ~log_size (Conf.root config)
```

Comment:
Isn't the log size irrelevant for a read-only instance? Is this just for consistency?

Reply:
`log_size` isn't an optional parameter

Reply:
Good point. (We should probably have had separate constructors for RO / RW from the start.)